### PR TITLE
IGNITE-23293 Implement the algorithm of Metastorage Compaction Trigger

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-23293

### CODE REVIEW

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java [hunk 1]
```diff
@@ -224,34 +214,24 @@ public long getCompactionRevision() {
...
-                                log.error(
-                                        "Metastore compaction failed: [compactionRevision={}]",
-                                        unwrapCause(throwable),
-                                        compactionRevision
-                                );
-                            }
-
-                            notifyCompleteCompactionLocally(this.compactionRevision, throwable);
-                        }, compactionExecutor);
+                watchProcessor.addTaskToWatchEventQueue(() -> {
```
Comment (ibessonov) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java:230]: I think we should use something less abstract, like "onCompactionSomething". Passing an arbitrary closure is a road to abuse. People will start doing all sorts of bad things.

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java [hunk 2]
```diff
@@ -224,34 +214,24 @@ public long getCompactionRevision() {
...
         rwLock.writeLock().lock();
 
         try {
             assertCompactionRevisionLessThanCurrent(compactionRevision, rev);
 
+            saveCompactionRevision(compactionRevision, context);
+
             if (isInRecoveryState()) {
-                this.compactionRevision = compactionRevision;
+                setCompactionRevision(compactionRevision);
```
Comment (ibessonov) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java:228]: I think this step can be done when we start compaction after recovery. So the real `if (isInRecoveryState()) {` here branch should be a no-op.

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java [hunk 3]
```diff
@@ -334,18 +314,13 @@ public void watchExact(byte[] key, long rev, WatchListener listener) {
...
 
     @Override
-    public CompletableFuture<Void> readOperationsFuture(long compactionRevisionExcluded) {
-        return readOperationForCompactionTracker.collect(compactionRevisionExcluded);
-    }
-
-    @Override
-    public void registerCompactionListener(CompactionListener listener) {
-        compactionListeners.add(listener);
+    public void registerCompactionRevisionUpdateListener(CompactionRevisionUpdateListener listener) {
```
Comment (ibessonov) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java:317]: Why don't you reuse `OnRevisionAppliedCallback`? It has a weird name, I get it, but it's a good place. There's no need for an additional list of listeners. Or am I wrong?

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java (hunk 1 above)
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java]: fix it

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java (hunk 3 above)
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java]: fix it

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java (hunk 2 above)
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java:243]: We discussed it in person, let's leave it like that.

**Review by ibessonov — APPROVED** [review]

**Pull request merged**